PolyType source generates invalid code related to events, methods, and interfaces

We may have a perfect storm of interfaces, nested types, and methods/events, to get the source generator to produce uncompilable code. 

## Repro steps

```
git clone https://github.com/microsoft/vs-streamjsonrpc.git
cd vs-streamjsonrpc
git checkout 69bf0deb58c7ae0071f179f5101495bd74130db7 # dev/andarno/eventShapes
.\StreamJsonRpc.slnx
```

Try to build.

```
1>------ Build started: Project: StreamJsonRpc.Tests, Configuration: Debug Any CPU ------
1>E:\src\vs-streamjsonrpc\obj\test\StreamJsonRpc.Tests\Debug\net472\generated\PolyType.SourceGenerator\PolyType.SourceGenerator.PolyTypeGenerator\PolyType.SourceGenerator.TypeShapeProvider.TargetObjectEventsTests_Server.g.cs(504,62,504,140): error CS0246: The type or namespace name '__s___EventAccessor_TargetObjectEventsTests_Server_add_TargetObjectEventsTests' could not be found (are you missing a using directive or an assembly reference?)
1>E:\src\vs-streamjsonrpc\obj\test\StreamJsonRpc.Tests\Debug\net472\generated\PolyType.SourceGenerator\PolyType.SourceGenerator.PolyTypeGenerator\PolyType.SourceGenerator.TypeShapeProvider.TargetObjectEventsTests_Server.g.cs(504,149,504,197): error CS0106: The modifier 'private' is not valid for this item
1>E:\src\vs-streamjsonrpc\obj\test\StreamJsonRpc.Tests\Debug\net472\generated\PolyType.SourceGenerator\PolyType.SourceGenerator.PolyTypeGenerator\PolyType.SourceGenerator.TypeShapeProvider.TargetObjectEventsTests_Server.g.cs(505,29,505,103): error CS0246: The type or namespace name '__EventAccessor_TargetObjectEventsTests_Server_add_TargetObjectEventsTests' could not be found (are you missing a using directive or an assembly reference?)
1>E:\src\vs-streamjsonrpc\obj\test\StreamJsonRpc.Tests\Debug\net472\generated\PolyType.SourceGenerator\PolyType.SourceGenerator.PolyTypeGenerator\PolyType.SourceGenerator.TypeShapeProvider.TargetObjectEventsTests_Server.g.cs(505,112,505,149): error CS0106: The modifier 'private' is not valid for this item
1>E:\src\vs-streamjsonrpc\obj\test\StreamJsonRpc.Tests\Debug\net472\generated\PolyType.SourceGenerator\PolyType.SourceGenerator.PolyTypeGenerator\PolyType.SourceGenerator.TypeShapeProvider.TargetObjectEventsTests_Server.g.cs(511,62,511,143): error CS0246: The type or namespace name '__s___EventAccessor_TargetObjectEventsTests_Server_remove_TargetObjectEventsTests' could not be found (are you missing a using directive or an assembly reference?)
1>E:\src\vs-streamjsonrpc\obj\test\StreamJsonRpc.Tests\Debug\net472\generated\PolyType.SourceGenerator\PolyType.SourceGenerator.PolyTypeGenerator\PolyType.SourceGenerator.TypeShapeProvider.TargetObjectEventsTests_Server.g.cs(511,152,511,200): error CS0106: The modifier 'private' is not valid for this item
1>E:\src\vs-streamjsonrpc\obj\test\StreamJsonRpc.Tests\Debug\net472\generated\PolyType.SourceGenerator\PolyType.SourceGenerator.PolyTypeGenerator\PolyType.SourceGenerator.TypeShapeProvider.TargetObjectEventsTests_Server.g.cs(512,29,512,106): error CS0246: The type or namespace name '__EventAccessor_TargetObjectEventsTests_Server_remove_TargetObjectEventsTests' could not be found (are you missing a using directive or an assembly reference?)
1>E:\src\vs-streamjsonrpc\obj\test\StreamJsonRpc.Tests\Debug\net472\generated\PolyType.SourceGenerator\PolyType.SourceGenerator.PolyTypeGenerator\PolyType.SourceGenerator.TypeShapeProvider.TargetObjectEventsTests_Server.g.cs(512,115,512,152): error CS0106: The modifier 'private' is not valid for this item
1>E:\src\vs-streamjsonrpc\obj\test\StreamJsonRpc.Tests\Debug\net472\generated\PolyType.SourceGenerator\PolyType.SourceGenerator.PolyTypeGenerator\PolyType.SourceGenerator.TypeShapeProvider.TargetObjectEventsTests_Server.g.cs(504,62,504,148): error CS0538: '__s___EventAccessor_TargetObjectEventsTests_Server_add_TargetObjectEventsTests.IServer' in explicit interface declaration is not an interface
1>E:\src\vs-streamjsonrpc\obj\test\StreamJsonRpc.Tests\Debug\net472\generated\PolyType.SourceGenerator\PolyType.SourceGenerator.PolyTypeGenerator\PolyType.SourceGenerator.TypeShapeProvider.TargetObjectEventsTests_Server.g.cs(504,149,504,197): error CS0501: 'TypeShapeProvider_StreamJsonRpc_Tests.ExplicitInterfaceImplementation_Event_MethodInfo()' must declare a body because it is not marked abstract, extern, or partial
1>E:\src\vs-streamjsonrpc\obj\test\StreamJsonRpc.Tests\Debug\net472\generated\PolyType.SourceGenerator\PolyType.SourceGenerator.PolyTypeGenerator\PolyType.SourceGenerator.TypeShapeProvider.TargetObjectEventsTests_Server.g.cs(505,29,505,111): error CS0538: '__EventAccessor_TargetObjectEventsTests_Server_add_TargetObjectEventsTests.IServer' in explicit interface declaration is not an interface
1>E:\src\vs-streamjsonrpc\obj\test\StreamJsonRpc.Tests\Debug\net472\generated\PolyType.SourceGenerator\PolyType.SourceGenerator.PolyTypeGenerator\PolyType.SourceGenerator.TypeShapeProvider.TargetObjectEventsTests_Server.g.cs(511,62,511,151): error CS0538: '__s___EventAccessor_TargetObjectEventsTests_Server_remove_TargetObjectEventsTests.IServer' in explicit interface declaration is not an interface
1>E:\src\vs-streamjsonrpc\obj\test\StreamJsonRpc.Tests\Debug\net472\generated\PolyType.SourceGenerator\PolyType.SourceGenerator.PolyTypeGenerator\PolyType.SourceGenerator.TypeShapeProvider.TargetObjectEventsTests_Server.g.cs(511,152,511,200): error CS0501: 'TypeShapeProvider_StreamJsonRpc_Tests.ExplicitInterfaceImplementation_Event_MethodInfo()' must declare a body because it is not marked abstract, extern, or partial
1>E:\src\vs-streamjsonrpc\obj\test\StreamJsonRpc.Tests\Debug\net472\generated\PolyType.SourceGenerator\PolyType.SourceGenerator.PolyTypeGenerator\PolyType.SourceGenerator.TypeShapeProvider.TargetObjectEventsTests_Server.g.cs(512,29,512,114): error CS0538: '__EventAccessor_TargetObjectEventsTests_Server_remove_TargetObjectEventsTests.IServer' in explicit interface declaration is not an interface
1>Done building project "StreamJsonRpc.Tests.csproj" -- FAILED.
```

Here is a sample of the code it generates that's invalid:

```cs
private static global::System.Reflection.MethodInfo? __s___EventAccessor_TargetObjectEventsTests_Server_add_TargetObjectEventsTests.IServer.ExplicitInterfaceImplementation_Event_MethodInfo;
private static void __EventAccessor_TargetObjectEventsTests_Server_add_TargetObjectEventsTests.IServer.ExplicitInterfaceImplementation_Event(global::TargetObjectEventsTests.Server? obj, global::System.EventHandler handler)
{
    global::System.Reflection.MethodInfo methodInfo = __s___EventAccessor_TargetObjectEventsTests_Server_add_TargetObjectEventsTests.IServer.ExplicitInterfaceImplementation_Event_MethodInfo ??= typeof(global::TargetObjectEventsTests.Server).GetEvent("TargetObjectEventsTests.IServer.ExplicitInterfaceImplementation_Event", __BindingFlags_All)!.AddMethod!;
    methodInfo.Invoke(obj!, new object?[] { handler });
}
```

--- a/src/PolyType.SourceGenerator/Parser/Parser.cs
+++ b/src/PolyType.SourceGenerator/Parser/Parser.cs
@@ -477,9 +477,21 @@ public sealed partial class Parser : TypeDataModelGenerator
         {
             customName = null;
 
+            if (!SyntaxFacts.IsValidIdentifier(eventSymbol.Name) ||
+                eventSymbol.IsImplicitlyDeclared ||
+                eventSymbol.HasAttribute(KnownSymbols.CompilerGeneratedAttribute))
+            {
+                return false; // Skip events that are explicit interface implementations or compiler-generated.
+            }
+
             if (ParseEventShapeAttribute(eventSymbol, out customName, out bool? ignore))
             {
                 return ignore is not true; // Skip events explicitly marked as ignored.
+            }
+
+            if (eventSymbol.DeclaredAccessibility is not Accessibility.Public)
+            {
+                return false; // Skip events that are not public when not annotated.
             }
 
             BindingFlags requiredFlags = eventSymbol.IsStatic ? BindingFlags.Public | BindingFlags.Static : BindingFlags.Public | BindingFlags.Instance;

--- a/src/PolyType/ReflectionProvider/ReflectionTypeShape.cs
+++ b/src/PolyType/ReflectionProvider/ReflectionTypeShape.cs
@@ -76,9 +76,9 @@ internal abstract class ReflectionTypeShape<T>(ReflectionTypeShapeProvider provi
                 return !shapeAttribute.Ignore; // Skip methods explicitly marked as ignored.
             }
 
-            if (!methodInfo.IsPublic || methodInfo.IsSpecialName)
+            if (!methodInfo.IsPublic)
             {
-                return false; // Skip methods that are not public or special names (like property getters/setters).
+                return false; // Skip methods that are not public and unannotated.
             }
 
             if (methodInfo.DeclaringType == typeof(object) || methodInfo.DeclaringType == typeof(ValueType))
@@ -110,6 +110,11 @@ internal abstract class ReflectionTypeShape<T>(ReflectionTypeShapeProvider provi
 
         bool IncludeEvent(EventInfo eventInfo, EventShapeAttribute? eventAttr)
         {
+            if (eventInfo.IsSpecialName || eventInfo.GetCustomAttribute<CompilerGeneratedAttribute>() is not null)
+            {
+                return false; // Skip events that are special names or compiler-generated.
+            }
+
             if (eventAttr is not null)
             {
                 return !eventAttr.Ignore;
@@ -118,7 +123,7 @@ internal abstract class ReflectionTypeShape<T>(ReflectionTypeShapeProvider provi
             MethodInfo? accessor = eventInfo.AddMethod ?? eventInfo.RemoveMethod;
             if (accessor is not { IsPublic: true })
             {
-                return false;
+                return false; // Skip events that are not public and unannotated.
             }
 
             MethodShapeFlags requiredFlag = accessor.IsStatic ? MethodShapeFlags.PublicStatic : MethodShapeFlags.PublicInstance;

--- a/tests/PolyType.SourceGenerator.UnitTests/CompilationTests.FunctionShapes.cs
+++ b/tests/PolyType.SourceGenerator.UnitTests/CompilationTests.FunctionShapes.cs
@@ -106,4 +106,45 @@ public static partial class CompilationTests
         PolyTypeSourceGeneratorResult result = CompilationHelpers.RunPolyTypeSourceGenerator(compilation);
         Assert.Empty(result.Diagnostics);
     }
+
+    [Fact]
+    public static void InterfaceMethods_ExplicitInterfaceImplemenentation_NoErrors()
+    {
+        // Regression test for https://github.com/eiriktsarpalis/PolyType/issues/293
+        Compilation compilation = CompilationHelpers.CreateCompilation("""
+            using System;
+            using PolyType;
+
+            #pragma warning disable CS0067 // Event is never used
+
+            public interface IServer
+            {
+                int Property { get; }
+                public int Method(int x, int y);
+                event Action<int>? InterfaceEvent;
+            }
+
+            public class Container
+            {
+
+                public class Server : IServer
+                {
+                    int IServer.Property => 42;
+                    event Action<int>? IServer.InterfaceEvent
+                    {
+                        add { _ = value; }
+                        remove { _ = value; }
+                    }
+
+                    int IServer.Method(int x, int y) => x + y;
+                }
+            }
+
+            [GenerateShapeFor(typeof(Container.Server), IncludeMethods = MethodShapeFlags.PublicInstance)]
+            public partial class Witness { }
+            """);
+
+        PolyTypeSourceGeneratorResult result = CompilationHelpers.RunPolyTypeSourceGenerator(compilation);
+        Assert.Empty(result.Diagnostics);
+    }
 }

